### PR TITLE
[Next.js] Performance improvements for editing integration

### PR DESCRIPTION
**Pull request by ambrauer**
## Description / Motivation
Improves performance for 'api/editing/render' Next.js API route when utilizing a self-hosted rendering/editing host (e.g. local or Sitecore XM Cloud) as opposed to a serverless deployment architectures (e.g. Vercel). We can greatly simplify the [`EditingDataService` architecture](https://doc.sitecore.com/xp/en/developers/hd/200/sitecore-headless-development/architecture-and-apis-for-integrating-jss-next-js-apps-with-sitecore-editors.html) in this case.

* `EditingDataService` is now an interface
* `ServerlessEditingDataService` & `BasicEditingDataService` implementations added
* by default, uses `ServerlessEditingDataService` when [Vercel is detected](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables:~:text=Description-,VERCEL,the%20app%20is%20deployed%20and%20running%20on%20Vercel.%20Example%3A%201.,-CI);  `BasicEditingDataService` otherwise
* bumped to next v12.2.5
* moved all editing-related types to new "editing" submodule
* moved all "edge" middlewares to "middleware" submodule

## Testing Details
New unit tests and manual testing locally and on Vercel.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/sitecore-jss-nextjs/src/editing/editing-data-cache.ts [hunk 1]
```diff
@@ -27,23 +27,23 @@ export class EditingDataDiskCache implements EditingDataCache {
...
-  get(key: string): EditingData | undefined {
+  async get(key: string): Promise<EditingData | undefined> {
     const entry = this.cache.get(key);
     if (!entry.isCached) {
       console.warn(`Editing data cache miss for key ${key} at ${this.cache.root}`);
       return undefined;
     }
-    const data = JSON.parse(entry.value);
     // Remove immediately to preserve disk-space
+    // NOTE: sync-disk-cache actually uses fs.unlink so this is async
```
Comment (illiakovalenko) [inline packages/sitecore-jss-nextjs/src/editing/editing-data-cache.ts:44]: It uses `unlinkSync` which is still sync:
https://github.com/stefanpenner/sync-disk-cache/blob/master/index.js#L229
https://github.com/stefanpenner/sync-disk-cache/blob/master/index.js#L10

File: packages/sitecore-jss-nextjs/src/editing/editing-data-cache.ts [hunk 2]
```diff
@@ -27,23 +27,23 @@ export class EditingDataDiskCache implements EditingDataCache {
     this.cache = new Cache('editing-data', { compression: 'gzip', location: tmpDir });
   }
 
-  set(key: string, editingData: EditingData): void {
+  async set(key: string, editingData: EditingData): Promise<void> {
```
Comment (illiakovalenko) [inline packages/sitecore-jss-nextjs/src/editing/editing-data-cache.ts:30]: Here is a sync operation, are there any changes?

**Review by illiakovalenko — CHANGES_REQUESTED** [review]
Looks very nice!
See some questions below, and other question:
As we discussed and decided yesterday, do we want to have `edge` module instead of `middleware`?

File: packages/sitecore-jss-nextjs/src/editing/editing-data-cache.ts (hunk 1 above)
Comment (ambrauer) [reply packages/sitecore-jss-nextjs/src/editing/editing-data-cache.ts:44]: 🤦‍♂️Yes you're absolutely right. I was looking at [async-disk-cache](https://github.com/stefanpenner/async-disk-cache/blob/master/index.js#L12). I had at one point switched over to this, but ran into some issues (webpack warnings, plus additional dependencies) that made it less attractive. I'll revisit...

**Comment by ambrauer** [comment]
> As we discussed and decided yesterday, do we want to have `edge` module instead of `middleware`?
I thought about this more and I think "middleware" is correct. This follows our naming as well as Next.js naming. We had discussed that it's possible we would have Next.js middleware that targets node runtime (instead of edge), but this isn't a valid for Next.js Middleware - only API Routes. We can safely assume anything in this submodule will target the edge runtime.

> What do you think about splitting `BasicEditingDataService` and `ServerlessEditingDataService` by files?
I had considered this too. I might do this, but I think it's also concise enough to keep them combined for now.

File: packages/sitecore-jss-nextjs/src/editing/editing-data-cache.ts (hunk 1 above)
Comment (ambrauer) [reply packages/sitecore-jss-nextjs/src/editing/editing-data-cache.ts]: @illiakovalenko See latest. Stuck with sync-disk-cache (and sync `EditingData[Disk]Cache`). Wrapped the cache remove in setTimeout so it's added to the macrotasks queue (non-blocking).

**Review by illiakovalenko — APPROVED** [review]
Looks awesome 👍

**Pull request merged**